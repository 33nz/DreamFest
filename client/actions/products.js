export const REQUEST_PRODUCTS = 'REQUEST_PRODUCTS'
export const RECEIVE_PRODUCTS = 'RECEIVE_PRODUCTS'

export const requestProducts = () => {
  return {
    type: REQUEST_PRODUCTS
  }
}

export const receiveProducts = products => {
  return {
    type: RECEIVE_PRODUCTS,
    products: products
  }
}
