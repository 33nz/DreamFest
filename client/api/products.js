import request from 'superagent'

const apiRoot = '/api/v1/'

export function getProducts () {
  return request.get(`${apiRoot}products`)
    .then(res => res.body)
}
