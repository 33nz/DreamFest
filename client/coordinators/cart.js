export function updateCartItem (updateInfo, updateCartDispatcher) {
  const isValid = !isNaN(Number(updateInfo.quantity))
  isValid && updateCartDispatcher(updateInfo)
}
