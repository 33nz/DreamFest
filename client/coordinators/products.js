import requestor from '../consume'

export function getProducts (dispatchers, consume = requestor) {
  const { fetchProductsPending, fetchProductsSuccess, showError } = dispatchers
  fetchProductsPending()
  return consume('/products')
    .then(res => {
      const products = res.body
      fetchProductsSuccess(products)
      return null
    })
    .catch(err => {
      showError(err.message)
    })
}
