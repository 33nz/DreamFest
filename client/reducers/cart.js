import {
  ADD_TO_CART,
  DELETE_FROM_CART,
  UPDATE_CART } from '../actions/cart'
import { POST_ORDER_SUCCESS } from '../actions/orders'
import { getNewCart } from './reducer-helpers'

const cart = (state = [], action) => {
  switch (action.type) {
    case ADD_TO_CART:
      return getNewCart(state, action.product)

    case DELETE_FROM_CART:
      return state.filter(item => item.id !== action.id)

    case UPDATE_CART:
      return action.cart

    case POST_ORDER_SUCCESS:
      return []

    default:
      return state
  }
}

export default cart
