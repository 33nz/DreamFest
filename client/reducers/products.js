import { RECEIVE_PRODUCTS } from '../actions/products'

const products = (state = [], action) => {
  switch (action.type) {
    case RECEIVE_PRODUCTS:
      return action.products

    default:
      return state
  }
}

export default products
