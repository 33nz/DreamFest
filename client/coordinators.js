import { showError } from './actions/error'
import { requestProducts, receiveProducts } from './actions/products'

import { fetchProducts } from './api'

export function getProducts (dispatch) {
  dispatch(requestProducts())
  fetchProducts()
    .then(products => {
      dispatch(receiveProducts(products))
    })
    .catch(err => {
      dispatch(showError(err.message))
    })
}
