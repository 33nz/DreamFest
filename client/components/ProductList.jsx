import React from 'react'
import { connect } from 'react-redux'

import { getProducts } from '../api/products'

import { addToCart } from '../actions/cart'
import { navigate } from '../actions/navigate'
import { showError } from '../actions/error'
import { requestProducts, receiveProducts } from '../actions/products'

import ProductListItem from './ProductListItem'
import WaitIndicator from './WaitIndicator'

class ProductList extends React.Component {
  componentDidMount () {
    const { dispatch } = this.props
    dispatch(requestProducts())
    getProducts()
      .then(products => {
        dispatch(receiveProducts(products))
      })
      .catch(err => {
        dispatch(showError(err.message))
      })
  }
  render () {
    const { products, dispatch } = this.props
    return (
      <div className='productlist'>
        <p className='welcome'>
          Welcome! Please choose from our delicious selection and don&apos;t
          hesitate to let us know if we can answer any of your questions.
        </p>
        <WaitIndicator />
        {products.map(product => {
          return (
            <ProductListItem
              key={product.id}
              product={product}
              addToCart={item => {
                dispatch(addToCart(item))
                dispatch(navigate('cart'))
              }}
            />
          )
        })}
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    products: state.products
  }
}

export default connect(
  mapStateToProps
)(ProductList)
