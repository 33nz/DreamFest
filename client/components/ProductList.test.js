import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import { ProductList } from './ProductList'

import mockProducts from '../testing/mockProducts'

import { getProducts } from '../coordinators/products'
import { addCartItem } from '../coordinators/cart'

jest.mock('../coordinators/products', () => {
  return {
    getProducts: jest.fn()
  }
})

jest.mock('../coordinators/cart', () => {
  return {
    addCartItem: jest.fn()
  }
})

test('renders a list of products correctly', () => {
  const mockChild = <>mock child component</>
  const { asFragment } = render(<ProductList products={mockProducts}>
    {mockChild}</ProductList>)
  expect(asFragment()).toMatchSnapshot()
})

test('calls getOrders on mount', () => {
  render(<ProductList products={mockProducts} />)
  expect(getProducts).toHaveBeenCalled()
})

test('calls addCartItem on add to cart button click', async () => {
  render(<ProductList products={[ mockProducts[0] ]} />)

  const addButton = await screen.getByRole('button', { name: 'Add to cart' })
  fireEvent.click(addButton)

  expect(addCartItem).toHaveBeenCalled()
})
