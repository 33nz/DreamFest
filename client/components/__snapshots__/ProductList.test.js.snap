// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a list of products correctly 1`] = `
<DocumentFragment>
  <div
    class="productlist"
  >
    <div
      class="welcome"
    >
      <p>
        Welcome! Please choose from our delicious selection and don't hesitate to let us know if we can answer any of your questions.
      </p>
    </div>
    mock child component
    <div
      class="product"
    >
      <p
        class="name"
      >
        mocked product 1
      </p>
      <p
        class="description"
      >
        snazzy product
      </p>
      <p>
        <span
          class="country"
        >
          Lovingly made in testlandia
        </span>
        <button
          class="cart-link"
        >
          Add to cart
        </button>
      </p>
    </div>
    <div
      class="product"
    >
      <p
        class="name"
      >
        mocked product 2
      </p>
      <p
        class="description"
      >
        snazzier product
      </p>
      <p>
        <span
          class="country"
        >
          Lovingly made in testlandia
        </span>
        <button
          class="cart-link"
        >
          Add to cart
        </button>
      </p>
    </div>
    <div
      class="product"
    >
      <p
        class="name"
      >
        mocked product 3
      </p>
      <p
        class="description"
      >
        snazziest product
      </p>
      <p>
        <span
          class="country"
        >
          Lovingly made in testlandia
        </span>
        <button
          class="cart-link"
        >
          Add to cart
        </button>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
